contributors/design-proposals/termination-reason.md

```diff
@@ -0,0 +1,345 @@
+with a declared `PreStop` handler in accordance with the configuration of its 
+`ReasonDelivery`.
+
+### Termination Reason Delivery
+
+When Kublet processes the `PreStop` Handlers of a Pod's containers, prior to 
+sending a `TERM` signal and starting the termination grace timer, if a 
+termination reason is to be delivered, Kubelet will do the following.
+
+1. If the `PreStop` handler indicates a command action, Kubelet will supply the 
+termination reason to the container based on the following criteria.
+   1. If the `ReasonDelivery` is nil Kubelet will set the 
+   `DefaultTerminationReasonEnv` to the value of the termination reason. This 
+   is the default method of delivery for a command action pre-stop handler.
+   1. If the `Env` field of the `TerminationReasonDelivery` is not nil, Kubelet 
+   will set the environment variable indicated by this field to the value of 
+   the termination reason.
+   1. If none of the above criteria are met, then the pre-stop handler is 
+   malformed, and API Server [validation](#validation) has failed to reject 
+   its creation or update. In this case, Kubelet will log an error and deliver 
+   the termination reason via the default method specified above.
+1. If the `PreStop` handler indicates a HTTP GET action, Kublet will supply the 
+termination reason to the specified endpoint based on the following criteria.
+   1. If the `ReasonDelivery` is nil Kubelet will set the header indicated by 
+   `DefaultTerminationReasonHeader` to the value of the termination reason. This 
+   is the default method of delivery for HTTP GET pre-stop handlers.
+   1. If `ReasonDelivery` is not nil, and if its `Header` field 
+   is not nil, Kubelet will add a header, whose name is indicated by the value 
+   of this field and whose value is the termination reason, to the HTTP GET 
+   request, prior to sending the request.
+   1. If none of the above criteria are met, then the pre-stop handler is 
+   malformed, and API Server [validation](#validation) has failed. In this case,
+   Kubelet will log an error and deliver the termination reason via the default
+   header specified above.
+
+### Pod Deletion 
+
+When Kubelet finds that a Pod's `DeletionTimestamp` is set, during its 
+termination processing, if a `TerminationReason` has been set, it will 
+[deliver the termination reason](#termination-reason-delivery).
+
+### Pod Eviction
+
+When Kubelet targets a Pod for eviction, it will 
+[deliver a termination reason](#termination-reason-delivery) of 
+`ReasonEviction`.
+
+### Intolerable Taints
+
+When Kubelet evicts a Pod due to an intolerable taint, it will 
+[delivery a termination reason](#termination-reason-delivery) of 
+`ReasonIntolerableTaint`.
+
+## Kubectl
+Kubectl will use the `--reason` parameter to allow users to pass a arbitrary 
+string as the termination reason as shown below.
+
+```shell
+ > kubectl delete po my-pod --reason="resolve issue 354961"
```

Comment:
...if that reason propagation behavior is desired, I think it should come in a phase two, because it's a lot of work.